### DISCUSSION TITLE
implementing dirty state to save changes in properties panel

### DISCUSSION THREAD

**Discussion by Adem-Kri** [eclipse-glsp/glsp#1144] [Q&A]
Hello,

I am currently working on a graph that has a properties panel which displays information about the selected element. This panel also allows me to make changes to the element's properties. 

However, I have noticed that when I hit save, the changes I made are not saved unless I move something in the graph. After some research, I found that setting the dirty state can help solve this issue. 

Could you please explain how to implement the dirty state in my project? 

Thank you.

![image](https://github.com/eclipse-glsp/glsp/assets/116633370/7d51763a-d411-4fc8-915c-bb1d1721dcdf)

**Comment by rsoika** [comment]
If you see that the dirty flag has changed on the client size this usually only indicates that the GLSP client Model has changed. This does not necessarily mean that your server model was updated too. At least this was a topic which took me a long time to understand in my own project. It is necessary that on each change operation you update your source model. But my assumption here is that your source model and your glsp model are not the same. If this is the case in your project than you can ignore my answer as I am not the expert here ;-)

**Reply by Adem-Kri** [reply]
Thank you for your answer @rsoika,
the dirty flag isn't changing when I insert modifications in the properties panel. the flag is only set when I move an element or add a new one in the graph.

**Reply by rsoika** [reply]
ok, this indicates that your property panel changes did not affect the GLSP model - so the dirty flag is not updated.
Have you tried to send the isDirty flag explicit as an Action event?

```java
// java sever code example:
 this.actionDispatcher.dispatch(
                            new SetDirtyStateAction(true, SetDirtyStateAction.Reason.OPERATION));
```

**Comment by tortmayr** [comment]
Hi @Adem-Kri,
>I am currently working on a graph that has a properties panel which displays information about the selected element. This panel also allows me to make changes to the element's properties.

How do you propagate this changes to the glsp server? The diry state is only updated automatically if you send (& handle) an Operation to the server. Plain actions typically should not modifiy the source model. If you can't (or don't want to) send an operation, then you should send a `SetDirtyAction` with `isDirty` set to `true` after each edit in the property panel.
